expression/simple_rewriter.go

```diff
@@ -0,0 +1,455 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"fmt"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/parser"
+	"github.com/pingcap/tidb/parser/opcode"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+)
+
+type simpleRewriter struct {
+	tbl *model.TableInfo
+
+	stack *ExprStack
+	err   error
+	ctx   sessionctx.Context
+}
+
+// ParseSimpleExpr parses simple expression string to Expression.
+// The expression string must only reference the column in table Info.
+func ParseSimpleExpr(ctx sessionctx.Context, exprStr string, tableInfo *model.TableInfo) (Expression, error) {
+	exprStr = fmt.Sprintf("select %s", exprStr)
+	stmts, err := parser.New().Parse(exprStr, "", "")
```

Comment:
Maybe we can pack parser, rewriter, string buffer and so on into a single struct, and provide ParseSimpleExpr as a method.
The parser is a bit heavy because it use cache for object allocation, but I'm not sure the use pattern of ParseSimpleExpr function so I don't know whether it's worth to do it.

Reply:
If it is used extensively, we can consider reusing the parser then.